### PR TITLE
Vector search c

### PR DESCRIPTION
**Pull request by gilkal**
Vector search test app handler for C. Tested with the vector search tests and other tests, to see that nothing is broken.

### CODE REVIEW

File: CBLClient/Apps/CBLTestServer-C/build_scripts/build_cross.py [hunk 1]
```diff
@@ -159,6 +167,12 @@ def check_sysroot(name: str):
         tar.extractall(DOWNLOAD_DIR)
     os.remove(zip_filename)
 
+    vector_search_zip_name = f'couchbase-lite-vector-search-1.0.0-42-linux-x86_64.zip'
```
Comment (stanton11) [inline CBLClient/Apps/CBLTestServer-C/build_scripts/build_cross.py:170]: Hardcoding VectorSearch version here, would it be worth leaving a comment in to suggest updating with future versions.

File: CBLClient/Apps/CBLTestServer-C/src/CollectionMethods.cpp [hunk 2]
```diff
@@ -150,6 +154,61 @@ namespace collection_methods {
...
+                    }
+                    else {
+                        FLDict docProperties = CBLDocument_Properties(document);
+                        FLMutableDict_SetValue(documents, FLValue_AsString(docId), reinterpret_cast<FLValue>(docProperties));
+                        CBLDocument_Release(document);
+                    }
+                }
+            });
+            write_serialized_body(conn, memory_map::store(documents, FLMutableDict_EntryDelete));
+        });*/
```
Comment (stanton11) [inline CBLClient/Apps/CBLTestServer-C/src/CollectionMethods.cpp:185]: commented code to be kept?

File: CBLClient/Apps/CBLTestServer-C/src/CollectionMethods.cpp [hunk 3]
```diff
@@ -415,9 +499,9 @@ namespace collection_methods {
...
 
         success = true;
-    });
-    });
+        });
+        });
 
-    write_empty_body(conn);
-}
+     write_empty_body(conn);
```
Comment (stanton11) [inline CBLClient/Apps/CBLTestServer-C/src/CollectionMethods.cpp:505]: leading whitespace

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+            else if (metric == "cosine")
+            {
+                dMetric = kCBLDistanceMetricCosine;
+            }
+            else
+            {
+                throw std::invalid_argument("Invalid distance metric");
+            }
+
+            // CBLVectorIndexConfiguration config{kCBLN1QLLanguage, expression, dimensions, centroids};
```
Comment (stanton11) [inline CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp:176]: commented code can be deleted

**Review by stanton11 — CHANGES_REQUESTED** [review]
Looks good, few comments

File: CBLClient/Apps/CBLTestServer-C/build_scripts/build_cross.py (hunk 1 above)
Comment (gilkal) [reply CBLClient/Apps/CBLTestServer-C/build_scripts/build_cross.py]: Thanks for picking that up, fixed! (tested with building and then running the vector. search test)

File: CBLClient/Apps/CBLTestServer-C/src/CollectionMethods.cpp (hunk 3 above)
Comment (gilkal) [reply CBLClient/Apps/CBLTestServer-C/src/CollectionMethods.cpp]: Thanks, fixed.

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp (hunk 4 above)
Comment (gilkal) [reply CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp]: Thanks, fixed.

File: CBLClient/Apps/CBLTestServer-C/src/CollectionMethods.cpp (hunk 2 above)
Comment (gilkal) [reply CBLClient/Apps/CBLTestServer-C/src/CollectionMethods.cpp]: Thanks, fixed

**Review by stanton11 — APPROVED** [review]
lgtm

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+            config.centroids = centroids;
+            config.minTrainingSize = minTrainingSize;
+            config.maxTrainingSize = maxTrainingSize;
+            config.expressionLanguage = kCBLN1QLLanguage;
+            with<CBLDatabase *>(body,"database", [conn, collectionName, scopeName, indexName, config](CBLDatabase* db)
+            {
+                CBLError err;
+                CBLCollection* collection;
+                TRY(collection = CBLDatabase_CreateCollection(db, flstr(collectionName),  flstr(scopeName), &err), err);
+                TRY(CBLCollection_CreateVectorIndex(collection, flstr(indexName), config, &err), err);
```
Comment (pasin) [inline CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp:190]: If the config.encoding is not nullptr, calling `CBLVectorEncoding_Free(config.encoding);`

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+                FLMutableDict_SetArray(words, FLValue_AsString(word), FLValue_AsArray(vector));
+            }
+         }
+         CBLQuery_Release(query);
+         CBLResultSet_Release(rs);
+         TRY(CBLDatabase_Close(db, &err), err);
+         return words;
+}
+
+FLDict getEmbeddingDic(string term) {
```
Comment (pasin) [inline CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp:97]: `Dict`?

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+        write_serialized_body(conn, "Model registered");
+    }
+    
+    void vectorSearch_getEmbedding(json& body, mg_connection* conn)
+    {
+        auto embbedingsVector =  FLMutableArray_New();
+        DEFER {
+            FLMutableArray_Release(embbedingsVector);
+        };
+        auto vectorDict = getEmbeddingDic(body["input"].get<string>());
```
Comment (pasin) [inline CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp:230]: Need to release vectorDict as well.

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        }
+        write_serialized_body(conn, (FLValue)embbedingsVector);
+    }
+
+
+    void vectorSearch_query(json& body, mg_connection* conn) {
+
+        with<CBLDatabase *>(body,"database", [conn, body](CBLDatabase* db)
+            {
+                auto embeddedTermDic = getEmbeddingDic(body["term"].get<string>());
```
Comment (pasin) [inline CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp:247]: Need to release embeddedTermDic as well.

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp (hunk 5 above)
Comment (gilkal) [reply CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp:190]: Thanks,  fixed

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp (hunk 7 above)
Comment (gilkal) [reply CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp]: Thanks,  fixed

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp (hunk 8 above)
Comment (gilkal) [reply CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp]: Thanks, fixed

File: CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp (hunk 6 above)
Comment (gilkal) [reply CBLClient/Apps/CBLTestServer-C/src/VectorSearchMethods.cpp]: Thanks, fixed

**Pull request merged**